DataFusion queries don't use the AWS SDK's built-in credential caching

### Description / Background

Error whilst running lots of parallel DataFusion queries.

### Steps to reproduce

Run lots of parallel DataFusion queries. See error: "[ERROR] sleeper_df/src/lib.rs:230 - query error External error: an error occurred while loading credentials"

### Expected behaviour

It should be possible to run lots of parallel DataFusion queries without seeing this problem.

### Technical Notes / Implementation Details

We believe this issue is due to our object store layer creating a new credentials provider chain with an empty cache on every query.

--- a/rust/objectstore_ext/Cargo.toml
+++ b/rust/objectstore_ext/Cargo.toml
@@ -38,6 +38,7 @@ log = { workspace = true }
 num-format = { workspace = true }
 object_store = { workspace = true, features = ["aws"] }
 thiserror = { workspace = true }
+tokio = { workspace = true }
 url = { workspace = true }
 
 [dev-dependencies]

--- a/rust/objectstore_ext/src/s3.rs
+++ b/rust/objectstore_ext/src/s3.rs
@@ -82,21 +82,19 @@ impl AwsConfig {
     /// provider is used.
     ///
     /// If a region is not specified, then the region is read from the default provider.
-    async fn to_builder(&self) -> color_eyre::Result<AmazonS3Builder> {
-        let config = sdk_credentials().await;
-
+    async fn to_builder(&self, config: &SdkConfig) -> color_eyre::Result<AmazonS3Builder> {
         let creds = match &self.secrets {
             Some(secrets) => Credentials::from_keys(
                 secrets.access_key_id.clone(),
                 secrets.secret_access_key.clone(),
                 secrets.session_token.clone(),
             ),
-            None => create_creds_from_config(&config).await?,
+            None => create_creds_from_config(config).await?,
         };
 
         let region = match &self.region {
             Some(region) => Region::new(region.clone()),
-            None => create_region_from_config(&config)?,
+            None => create_region_from_config(config)?,
         };
 
         Ok(builder_from_creds(&creds, &region, self.endpoint.as_ref()))
@@ -125,9 +123,21 @@ fn builder_from_creds(
     }
 }
 
-/// Load AWS configuration from the default providers.
-async fn sdk_credentials() -> SdkConfig {
-    aws_config::defaults(BehaviorVersion::latest()).load().await
+/// Caches a resolved [`SdkConfig`] so the default AWS credentials provider chain is only
+/// ever loaded once, rather than on every call. Loading it once and reusing it lets the
+/// AWS SDK's own credential caching and refresh-before-expiry logic deal with expiring
+/// credentials.
+#[derive(Debug, Default)]
+pub struct SdkConfigCache(tokio::sync::OnceCell<SdkConfig>);
+
+impl SdkConfigCache {
+    /// Return the cached [`SdkConfig`], loading it from the default AWS providers the
+    /// first time this is called.
+    pub async fn get(&self) -> &SdkConfig {
+        self.0
+            .get_or_init(|| async { aws_config::defaults(BehaviorVersion::latest()).load().await })
+            .await
+    }
 }
 
 /// Extract AWS credentials from an SDK configuration.
@@ -187,15 +197,17 @@ pub struct ObjectStoreFactory {
     s3_config: Option<AwsConfig>,
     store_map: RefCell<HashMap<String, Arc<dyn ObjectStore>>>,
     use_readahead: bool,
+    sdk_config: SdkConfig,
 }
 
 impl ObjectStoreFactory {
     #[must_use]
-    pub fn new(s3_config: Option<AwsConfig>, use_readahead: bool) -> Self {
+    pub fn new(s3_config: Option<AwsConfig>, use_readahead: bool, sdk_config: SdkConfig) -> Self {
         Self {
             s3_config,
             store_map: RefCell::new(HashMap::new()),
             use_readahead,
+            sdk_config,
         }
     }
 
@@ -310,7 +322,7 @@ impl ObjectStoreFactory {
             Some(config) => config,
             None => &AwsConfig::default(),
         };
-        cfg.to_builder().await
+        cfg.to_builder(&self.sdk_config).await
     }
 
     fn wrap_s3_store(

--- a/rust/sleeper_core/src/common_config.rs
+++ b/rust/sleeper_core/src/common_config.rs
@@ -16,8 +16,8 @@
 use crate::{
     datafusion::{OutputType, SleeperRegion},
     filter_aggregation_config::{aggregate::Aggregate, filter::Filter},
-    sleeper_context::SleeperContext,
 };
+use aws_config::SdkConfig;
 use color_eyre::eyre::{Result, bail};
 use objectstore_ext::s3::{AwsConfig, ObjectStoreFactory};
 use std::fmt::{Display, Formatter};
@@ -69,11 +69,7 @@ impl CommonConfig<'_> {
         self.sorting_columns_iter().collect::<Vec<_>>()
     }
 
-    pub(crate) async fn create_object_store_factory(
-        &self,
-        sleeper_context: &SleeperContext,
-    ) -> ObjectStoreFactory {
-        let sdk_config = sleeper_context.resolve_sdk_config().await.clone();
+    pub(crate) fn create_object_store_factory(&self, sdk_config: SdkConfig) -> ObjectStoreFactory {
         ObjectStoreFactory::new(
             self.aws_config.clone(),
             self.use_readahead_store,

--- a/rust/sleeper_core/src/common_config.rs
+++ b/rust/sleeper_core/src/common_config.rs
@@ -16,6 +16,7 @@
 use crate::{
     datafusion::{OutputType, SleeperRegion},
     filter_aggregation_config::{aggregate::Aggregate, filter::Filter},
+    sleeper_context::SleeperContext,
 };
 use color_eyre::eyre::{Result, bail};
 use objectstore_ext::s3::{AwsConfig, ObjectStoreFactory};
@@ -68,8 +69,16 @@ impl CommonConfig<'_> {
         self.sorting_columns_iter().collect::<Vec<_>>()
     }
 
-    pub(crate) fn create_object_store_factory(&self) -> ObjectStoreFactory {
-        ObjectStoreFactory::new(self.aws_config.clone(), self.use_readahead_store)
+    pub(crate) async fn create_object_store_factory(
+        &self,
+        sleeper_context: &SleeperContext,
+    ) -> ObjectStoreFactory {
+        let sdk_config = sleeper_context.resolve_sdk_config().await.clone();
+        ObjectStoreFactory::new(
+            self.aws_config.clone(),
+            self.use_readahead_store,
+            sdk_config,
+        )
     }
 
     pub(crate) fn output(&self) -> &OutputType {

--- a/rust/sleeper_core/src/lib.rs
+++ b/rust/sleeper_core/src/lib.rs
@@ -91,7 +91,8 @@ pub async fn run_compaction(
     sleeper_context: &Arc<SleeperContext>,
     progress_callback: Option<ProgressCallback>,
 ) -> Result<CompactionResult> {
-    let store_factory = config.create_object_store_factory(sleeper_context).await;
+    let sdk_config = sleeper_context.resolve_sdk_config().await.clone();
+    let store_factory = config.create_object_store_factory(sdk_config);
     let token = CancellationToken::new();
     let mut handle = None;
 
@@ -190,10 +191,8 @@ pub async fn run_query(
     config: &LeafPartitionQueryConfig<'_>,
     sleeper_context: &SleeperContext,
 ) -> Result<CompletedOutput> {
-    let store_factory = config
-        .common
-        .create_object_store_factory(sleeper_context)
-        .await;
+    let sdk_config = sleeper_context.resolve_sdk_config().await.clone();
+    let store_factory = config.common.create_object_store_factory(sdk_config);
     let rt = sleeper_context.retrieve_runtime_env()?;
 
     LeafPartitionQuery::new(config, &store_factory, rt)

--- a/rust/sleeper_core/src/lib.rs
+++ b/rust/sleeper_core/src/lib.rs
@@ -91,7 +91,7 @@ pub async fn run_compaction(
     sleeper_context: &Arc<SleeperContext>,
     progress_callback: Option<ProgressCallback>,
 ) -> Result<CompactionResult> {
-    let store_factory = config.create_object_store_factory();
+    let store_factory = config.create_object_store_factory(sleeper_context).await;
     let token = CancellationToken::new();
     let mut handle = None;
 
@@ -190,7 +190,10 @@ pub async fn run_query(
     config: &LeafPartitionQueryConfig<'_>,
     sleeper_context: &SleeperContext,
 ) -> Result<CompletedOutput> {
-    let store_factory = config.common.create_object_store_factory();
+    let store_factory = config
+        .common
+        .create_object_store_factory(sleeper_context)
+        .await;
     let rt = sleeper_context.retrieve_runtime_env()?;
 
     LeafPartitionQuery::new(config, &store_factory, rt)

--- a/rust/sleeper_core/src/sleeper_context.rs
+++ b/rust/sleeper_core/src/sleeper_context.rs
@@ -15,6 +15,7 @@
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
+use aws_config::SdkConfig;
 #[cfg(doc)]
 use datafusion::execution::object_store::ObjectStoreRegistry;
 use datafusion::{
@@ -23,6 +24,7 @@ use datafusion::{
     execution::runtime_env::{RuntimeEnv, RuntimeEnvBuilder},
     physical_plan::{ExecutionPlan, metrics::MetricsSet},
 };
+use objectstore_ext::s3::SdkConfigCache;
 use std::sync::{Arc, Mutex, Weak};
 
 /// A thread-safe class containing internal `DataFusion` context that needs to be
@@ -33,6 +35,10 @@ pub struct SleeperContext {
     inner: Mutex<RuntimeEnv>,
     // Weak pointers to currently executing compactions
     compaction_filter_stages: Mutex<HashMap<String, Weak<dyn ExecutionPlan>>>,
+    // Cached, lazily resolved default AWS SDK configuration, shared across all
+    // uses of this context so the credentials provider chain is only ever loaded
+    // once.
+    sdk_config_cache: SdkConfigCache,
 }
 
 /// The maximum size of `DataFusion`'s file metadata cache.
@@ -59,6 +65,13 @@ impl SleeperContext {
                 .with_metadata_cache_limit(METADATA_CACHE_SIZE)
                 .build()?,
         ))
+    }
+
+    /// Retrieves the resolved default AWS SDK configuration for this context, loading it
+    /// from the default credentials provider chain the first time it's needed and caching
+    /// it for the lifetime of this context.
+    pub async fn resolve_sdk_config(&self) -> &SdkConfig {
+        self.sdk_config_cache.get().await
     }
 
     /// Retrieves number of compaction rows read for given compaction job.

--- a/rust/sleeper_core/test_util/src/lib.rs
+++ b/rust/sleeper_core/test_util/src/lib.rs
@@ -35,7 +35,7 @@ use datafusion::{
 };
 use futures::StreamExt;
 use object_store::ObjectStoreExt;
-use objectstore_ext::s3::ObjectStoreFactory;
+use objectstore_ext::s3::{ObjectStoreFactory, SdkConfigCache};
 use rust_sketch::quantiles::{byte::byte_deserialize, i64::i64_deserialize, str::str_deserialize};
 use sleeper_core::{ColRange, DataSketchVariant, PartitionBound};
 use std::{collections::HashMap, fs::File, sync::Arc};
@@ -133,7 +133,8 @@ async fn deserialise_sketches(
     path: &Url,
     key_types: Vec<DataType>,
 ) -> color_eyre::Result<Vec<DataSketchVariant>> {
-    let factory = ObjectStoreFactory::new(None, true);
+    let sdk_config = SdkConfigCache::default().get().await.clone();
+    let factory = ObjectStoreFactory::new(None, true, sdk_config);
     deserialise_sketches_with_factory(&factory, path, key_types).await
 }
 

